Total ordering unit test

jpy-git opened the pull request:
<!--
Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

- [ ] Add a ChangeLog entry describing what your PR does.
- [ ] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [ ] Write a good description on what the PR does.
- [ ] If you used multiple emails or multiple names when contributing, add your mails
   and preferred name in ``script/.contributors_aliases.json``
-->

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

<!--
If this PR fixes a particular issue, use the following to automatically close that issue
once this PR gets merged:
-->

Closes #6045. Adds a unit test for the `__gt__` method mentioned in the linked issue.

N.B. since we don't currently implement a `__eq__` method (as described in the [docs](https://docs.python.org/3/library/functools.html#functools.total_ordering)) we can't currently assert something like `fake_checker_1 == fake_checker_2` even though they have the same name and args.

Didn't add it in since we're only really doing this for coverage and seems unnecessary, but happy to add in for completeness if desired.

DanielNoord commented on tests/checkers/unittest_base_checker.py:
```diff
@@ -65,3 +76,11 @@ def test_base_checker_doc() -> None:
     assert repr(less_basic) == repr(basic)
+
+
+def test_base_checker_ordering():
```
```suggestion
def test_base_checker_ordering() -> None:
    """Test ordering of checkers based on their __gt__ method."""
```

Pierre-Sassoulas reviewed (changes requested):
Let's add the ``__eq__`` we're taking some time for this may as well be exhaustive :)

jpy-git commented:
> Let's add the `__eq__` we're taking some time for this may as well be exhaustive :)

I've added in `__eq__` and a test for it 👍 
N.B. since we've added an `__eq__`, I've also had to add the corresponding `__hash__` to make the checkers hashable ([docs](https://docs.python.org/3/reference/datamodel.html#object.__hash__))
